### PR TITLE
Nightly cronjob CI

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes https://github.com/unicode-org/icu4x/issues/2718

This is reviewable now, but please do not merge yet, since for the purposes of testing I've made it so that the nightly CI is run for the pull_request task as well.

This adds a cronjob (and a manually-triggerable) CI task that runs everything[^1] under the latest nightly. It does this by conditionally setting env vars, so future changes to CI will automatically inherit this as long as new jobs use the same boilerplate. There are also env vars that allow us to skip or fixup tests that are testing things on specific compiler versions.

 [^1]: Except for the clippy build and the rustdoc-json using stuff

### CODE REVIEW

**Comment by Manishearth** [comment]
The FFI failure is https://github.com/rust-lang/rust/pull/103193

File: .github/workflows/build-test.yml
Comment (echeran) [inline .github/workflows/build-test.yml:54]: suggestion: replace the body of this repeated boilerplate with a script and a 1-liner invocation

According to the docs for [Github Actions environment variables](https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables), `GITHUB_EVENT_NAME` is an environment variable that is available to scripts already.  Even though that doesn't cover `inputs.nightly`, you can still pass that value as an argument to such a script, [as in this example](https://docs.github.com/en/actions/learn-github-actions/contexts#example-usage-of-the-inputs-context-in-a-reusable-workflow).

I'm imagining that such a script would allow you to replace:
```
if: ...
run: |
  ...
```
with just
```
run: cargo make maybe-set-nightly  --inputsNightly=${{ inputs.nightly }}
```

Edit: now there's a second step because there's an orthogonal dimension of the runner OS.  The script + one-liner can still work if you you add an extra CLI arg:

```
run: cargo make maybe-set-nightly  --inputsNightly=${{ inputs.nightly }} --runnerOS=${{ runner.os }}
```

Comment (Manishearth) [reply .github/workflows/build-test.yml:54]: Oh, that's a good call, yeah

Comment (Manishearth) [reply .github/workflows/build-test.yml:54]: The runner OS is not relevant if we use a script, it's only relevant when using the actions yaml stuff

Comment (Manishearth) [reply .github/workflows/build-test.yml:54]: I think I won't change the `if` condition, that seems highly Actions-specific, AND i like that when it doesn't get run the GHA UI shows it greyed out.

I'll just swap out the body.

Comment (Manishearth) [reply .github/workflows/build-test.yml:54]: Actually, the annoying thing is that ideally this gets run before cargo-make needs to be installed. Hmm.

Comment (Manishearth) [reply .github/workflows/build-test.yml]: So the tradeoff is that we must install cargo-make (usually cached) every time we need this. Unfortunate but ... fine I guess?

File: Makefile.toml
Comment (echeran) [inline Makefile.toml]: question: does this file whose path is in `GITHUB_ENV` persist across the workflow?

I wonder if we can run the boilerplate step in one job in order to initialize the values across jobs. Then have all other jobs depend on the initialization job.

Comment (Manishearth) [reply Makefile.toml]: I don't think so.

You should experiment with this later, I don't want to spend too much time cleaning this up. This is _far_ from the only instance of boilerplate we have.

File: .github/workflows/build-test.yml
Comment (Manishearth) [reply .github/workflows/build-test.yml]: It's _really_ annoying to get the clippy job working.

I'm not convinced this is a net improvement. The GHA files have tons of boilerplate already, perhaps we should write our own Actions package to reduce that, but until we do that it feels like we're not optimizing much, and both cargo-make and Actions are super temperamental so moving it over isn't necessarily a win.

Comment (Manishearth) [reply .github/workflows/build-test.yml]: Oh i realized I don't need to try to get clippy working.

**Comment by Manishearth** [comment]
@echeran I think it's ready for review again

To land, I need to disable the wearos task on nightly, or alternatively wait a couple days for the cortex fix to land upstream, and then I need to remove pull_request from the conditions.

**Comment by Manishearth** [comment]
Once this lands I also want to set up a notification for build failures so I can keep it working

**Review by echeran — DISMISSED** [review]
LGTM after a valiant slog through CI. Looking forward to seeing it in action.  I'll let you carefully decide when the right time to try to merge this should be.

**Comment by Manishearth** [comment]
I decided to let this CI be failing for the first few rounds. The upstream Rust fix has already landed, so it'll be in by the next nightly.

I haven't yet hooked up notifications or anything so it doesn't matter.

**Review by echeran — APPROVED** [review]

**Pull request merged**
